### DISCUSSION TITLE
How to call Ava runner/reporter to get test stats.

### DISCUSSION THREAD

**Discussion by gouravjeet** [avajs/ava#2787] [Q&A] [answered]
I am trying to use AVA in our test suite for e2e testing. Can somebody share an example on how to get test stats in js code like this?

Here is the pseudo code of what i am trying to achieve:

```
export async function runTests() {
  // Create Ava reporter
  const ava = new Ava({ timeout: 150_000 });
  const filenames = readdirSync(__dirname)
    .filter((s) => /\.test\.js$/.test(s))
    .map((filename) => path.join(__dirname, filename));
  
  // Load all test files
  for (const filename of filenames) {
    ava.addFile(filename);
  }

  // Get Stats
  const stats = await new Promise((resolve) => {
    const runner = ava.run(() => resolve(runner.stats));
  });

  ava.unloadFiles();

  return stats;  
}
```

I am looking for following stats: totalTestCount, passedTestCount, skippedTestCount, failedTestCount, testDuration

**Comment by novemberborn** [comment]
Other than through what's emitted by the TAP reporter, this isn't available in any structured way.

Long term we'd want this to be available through reporters, so you'd run AVA in a child process and handle the output.

**Reply by gouravjeet** [reply]
Thank you for quick response.

**Reply by anthumchris** [reply]
I'm happy to contribute to this so we can use it.  @novemberborn Are you willing to point me in the right direction of which file to start with?

**Reply by novemberborn** [reply]
Assuming TAP can't do this, then we need some other structured output. There's a lot of tech debt in our reporter infrastructure, so implementing another standard (if it exists) or something more custom will be pretty difficult at this point.

**Comment by richardeschloss** [comment, marked as answer]
A very simple concrete example in case one else lands on this page:

```js
import { spawn } from 'child_process'
import reporter from 'tap-json'

function runTest (file) {
  const ava = spawn('npx', [
    'ava',
    file
  ])
  ava.stdout
    .pipe(reporter())
    .on('data', (report) => {
      console.log(report)
      /* Output */
      /* {
        stats: { asserts: 1, passes: 1, failures: 0 },
        asserts: [
          {
            number: 1,
            comment: null,
            name: 'Module: adds plugin',
            ok: true,
            extra: {}
          }
        ]
      } */
    })
}

runTest('test/module.js')
```

Seems like TAP is the way to go because of all the reporters it already supports.
